kasa/discover.py

```diff
@@ -399,5 +408,95 @@ def _get_device_class(info: dict) -> Type[SmartDevice]:
                 return SmartLightStrip
 
             return SmartBulb
+        raise UnsupportedDeviceException("Unknown device type: %s" % type_)
+
+    @staticmethod
+    def _get_device_instance_legacy(data: bytes, ip: str, port: int) -> SmartDevice:
+        """Get SmartDevice from legacy 9999 response."""
+        try:
+            info = json_loads(TPLinkSmartHomeProtocol.decrypt(data))
+        except Exception as ex:
+            raise SmartDeviceException(
+                f"Unable to read response from device: {ip}: {ex}"
+            ) from ex
+
+        _LOGGER.debug("[DISCOVERY] %s << %s", ip, info)
+
+        device_class = Discover._get_device_class(info)
+        device = device_class(ip, port=port)
+        device.update_from_discover_info(info)
+        return device
 
-        raise SmartDeviceException("Unknown device type: %s" % type_)
+    @staticmethod
+    def _get_device_instance(
+        data: bytes, ip: str, port: int, credentials: Credentials
+    ) -> SmartDevice:
+        """Get SmartDevice from the new 20002 response."""
+        try:
+            info = json_loads(data[16:])
+            discovery_result = DiscoveryResult(**info["result"])
+        except Exception as ex:
+            raise UnsupportedDeviceException(
+                f"Unable to read response from device: {ip}: {ex}"
+            ) from ex
+
+        if (
+            discovery_result.mgt_encrypt_schm.encrypt_type == "KLAP"
+            and discovery_result.mgt_encrypt_schm.lv is None
+        ):
+            type_ = discovery_result.device_type
+            device_class = None
+            if type_.upper() == "IOT.SMARTPLUGSWITCH":
+                device_class = SmartPlug
+
+            if device_class:
+                _LOGGER.debug("[DISCOVERY] %s << %s", ip, info)
+                device = device_class(ip, port=port, credentials=credentials)
+                device.update_from_discover_info(discovery_result.get_dict())
+                device.protocol = TPLinkKlap(ip, credentials)
+                return device
+            else:
+                raise UnsupportedDeviceException(
+                    f"Unsupported device {ip} of type {type_}: {info}"
+                )
+        else:
+            raise UnsupportedDeviceException(f"Unsupported device {ip}: {info}")
+
+
+class DiscoveryResult(BaseModel):
+    """Base model for discovery result."""
+
+    class Config:
+        """Class for configuring model behaviour."""
+
+        allow_population_by_field_name = True
+
+    class MgtEncryptSchm(BaseModel):
+        """Base model for encryption scheme of discovery result."""
+
+        is_support_https: Optional[bool] = None
+        encrypt_type: Optional[str] = None
+        http_port: Optional[int] = None
+        lv: Optional[int] = None
+
+    device_type: str = Field(alias="device_type_text")
+    device_model: str = Field(alias="model")
+    ip: str = Field(alias="alias")
+    mac: str
+    mgt_encrypt_schm: MgtEncryptSchm
```

Comment:
```suggestion
    encryption_schema: EncryptionSchema = Field(alias="mgt_encrypt_schm")
```
I think this should work?